roboteam_ai/src/skills/AvoidBallForBallPlacement.cpp

```diff
@@ -9,80 +9,67 @@
-    auto lineDirection = ballPos - ballPlacementTargetLocation;
-    auto inversedLineDirectionX = Vector2(-lineDirection.x, lineDirection.y).stretchToLength(.5);
-    auto inversedLineDirectionY = Vector2(lineDirection.x, -lineDirection.y).stretchToLength(.5);
 
+AvoidBallForBallPlacement::AvoidBallForBallPlacement(std::string name, bt::Blackboard::Ptr blackboard)
+: Skill(std::move(name), std::move(blackboard)) { }
 
-    auto point = projectionOnLinePoint + inversedLineDirectionX;
-    if (Field::pointIsInField(point) && !positionIsTooCloseToLine(point)) {
-        targetToMoveTo = point;
-        return;
-    }
+bt::Node::Status AvoidBallForBallPlacement::onUpdate() {
+    auto robotPos = rtt::Vector2(robot->pos);
+    Vector2 force = {0, 0};
 
-    point = projectionOnLinePoint + inversedLineDirectionY;
-    if (Field::pointIsInField(point) && !positionIsTooCloseToLine(point)) {
-        targetToMoveTo = point;
-        return;
-    }
+    // forces from robots
+    for (auto otherRobot : World::getAllRobots()) {
+        double distance = robotPos.dist(otherRobot.pos);
+        Vector2 distanceVector = robotPos - otherRobot.pos;
 
-    // if we reach this point we are in a corner of the field.
-    inversedLineDirectionX.stretchToLength(2);
-    targetToMoveTo = control::ControlUtils::projectPositionToWithinField(projectionOnLinePoint + inversedLineDirectionX);
-    if (positionIsTooCloseToLine(targetToMoveTo)) {
-        inversedLineDirectionY.stretchToLength(2);
-        targetToMoveTo = control::ControlUtils::projectPositionToWithinField(projectionOnLinePoint + inversedLineDirectionY);
+        if (otherRobot.id != robot->id && distance < minRobotDistanceForForce) {
+            force = force + distanceVector.stretchToLength(1) * (robotWeight/(distance * distance));
+        }
     }
-}
 
-bt::Node::Status AvoidBallForBallPlacement::onUpdate() {
-    if (currentProgress == DONE) {
-      //  return bt::Node::Status::Success;
+    // check forces from ball
+    double distance = robotPos.dist(ball->pos);
+    if (distance < minBallDistanceForForce) {
+        Vector2 distanceVector = robotPos-ball->pos;
+        force = force+distanceVector.stretchToLength(1)*(ballWeight/(distance*distance));
     }
 
-    auto velocities = gtp.goToPos(robot, targetToMoveTo, control::GoToType::luTh);
-    auto robotPos = rtt::Vector2(robot->pos);
-
-    roboteam_msgs::RobotCommand cmd;
-    cmd.id = robot->id;
-    cmd.x_vel = static_cast<float>(velocities.x);
-    cmd.y_vel = static_cast<float>(velocities.y);
-    cmd.use_angle = 1;
-    cmd.w = static_cast<float>((targetToMoveTo - robot->pos).angle());
-    publishRobotCommand(cmd);
-
-    if (robotPos.dist(targetToMoveTo) > constants::GOTOPOS_LUTH_ERROR_MARGIN) {
-        // return bt::Node::Status::Success;
+    // forces from walls
+    double boundWidth =  Field::get_field().boundary_width;
+    double halfFieldLength = Field::get_field().field_length/2;
```

Comment:
you are getting the field three times here, it's probably better to do it like this:

field = Field::get_field();
halfFieldLength = field.field_length*0.5;
..
..